### PR TITLE
Feat/api 577 delete state

### PR DESCRIPTION
**Pull request by khan08**
API-577 and API-578

### CODE REVIEW

File: sheepdog/transactions/deletion/transaction.py [hunk 1]
```diff
@@ -177,3 +182,12 @@ def get_nodes(self, ids):
                 } for c in node._related_cases_from_cache]
                 for node in nodes
             }
+
+    def mark_indexd_delete(self, ids):
```
Comment (kulgan) [inline sheepdog/transactions/deletion/transaction.py]: I think you should at least retry once and then raise an exception if failure continues. Retry within the loop, so as to ensure one failure does not override all the hard work done so far

Comment (khan08) [reply sheepdog/transactions/deletion/transaction.py]: For error handling aspect, this is after nodes are committed.  A failure here will have some or all indexd doc not have deleted flag, while corresponding nodes are deleted. This is impossible to handle given indexd has no transaction support.

I see your point where we add retries, simply due to the unreliable nature of indexd.

File: sheepdog/utils/__init__.py [hunk 2]
```diff
@@ -41,6 +42,7 @@
     submitted_state(),
     UPLOADING_STATE,
     UPLOADED_STATE,
+    DELETE_STATE
```
Comment (anmaxvl) [inline sheepdog/utils/__init__.py]: Is this a desired behavior? i.e. a file node has been deleted, but we still allow file upload?

Comment (khan08) [reply sheepdog/utils/__init__.py]: Yes, it's basically a reupload.

File: sheepdog/transactions/deletion/entity.py [hunk 3]
```diff
@@ -10,7 +11,7 @@
 )
 
 
-ALLOWED_DELETION_FILE_STATES = [submitted_state()]
+ALLOWED_DELETION_FILE_STATES = [submitted_state(), DELETE_STATE]
```
[the remark this answers is not in the document]
Comment (khan08) [reply sheepdog/transactions/deletion/entity.py]: We should put a lot of state definition in gdcdictionary, but it's hardcoded everywhere from portal to sheepdog to ancillary service. I guess submitted_state was an effort to use gdcdictionary, but later on it other stuff went all hardcoded.

**Pull request merged**